yew-macro/src/html_tree/html_component.rs

```diff
@@ -86,90 +95,36 @@ impl ToTokens for HtmlComponent {
-            // using span of type because the error message goes something like "children method not found".
-            // If we could control the message, it should say "component X doesn't accept children" and point at the children.
-            quote_spanned! {ty.span()=>
-                .children(::yew::html::ChildrenRenderer::new(#children))
-            }
+        let props_ty = quote_spanned!(ty.span()=> <#ty as ::yew::html::Component>::Properties);
+        let children_renderer = if children.is_empty() {
+            None
         } else {
-            quote! {}
-        };
-
-        let init_props = match &props.prop_type {
-            PropType::List(list_props) => {
-                let set_props = list_props.iter().map(|HtmlProp { label, value, .. }| {
-                    quote_spanned! {value.span()=> .#label(
-                        #[allow(unused_braces)]
-                        <::yew::virtual_dom::VComp as ::yew::virtual_dom::Transformer<_, _>>::transform(
-                            #value
-                        )
-                    )}
-                });
-
-                quote_spanned! {ty.span()=>
-                    <<#ty as ::yew::html::Component>::Properties as ::yew::html::Properties>::builder()
-                        #(#set_props)*
-                        #set_children
-                        .build()
-                }
-            }
-            PropType::With(props) => {
-                quote! { #props }
-            }
-            PropType::None => quote_spanned! {ty.span()=>
-                <<#ty as ::yew::html::Component>::Properties as ::yew::html::Properties>::builder()
-                    #set_children
-                    .build()
-            },
+            Some(quote! { ::yew::html::ChildrenRenderer::new(#children) })
         };
+        let build_props = props.build_properties_tokens(&props_ty, children_renderer);
 
-        let node_ref = if let Some(node_ref) = &props.node_ref {
-            quote_spanned! {node_ref.span()=> #node_ref }
+        let special_props = props.special();
+        let node_ref = if let Some(node_ref) = &special_props.node_ref {
+            let value = &node_ref.value;
+            quote_spanned! {value.span()=> #value }
         } else {
             quote! { ::yew::html::NodeRef::default() }
         };
 
-        let key = if let Some(key) = &props.key {
-            quote_spanned! {key.span()=>
+        let key = if let Some(key) = &special_props.key {
+            let value = &key.value;
+            quote_spanned! {value.span()=>
                 #[allow(clippy::useless_conversion)]
-                Some(::std::convert::Into::<::yew::virtual_dom::Key>::into(#key))
+                Some(::std::convert::Into::<::yew::virtual_dom::Key>::into(#value))
             }
         } else {
-            quote! {None}
+            quote! { None }
         };
 
         tokens.extend(quote_spanned! {ty.span()=>
             {
```

Comment:
nit: might be able to drop these braces now

Reply:
You know what? I thought exactly the same thing, but when I made the change it caused a bunch of errors so I promptly reverted and then forgot about it.
I'll try again, this time with more resolve :smile:

Reply:
Ah I see what's happening.
When using the `html_nested!` macro you can end up with this:
```rust
let my_closure = || html_nested! { <Foo /> };

// turns into
let my_closure = || #[allow(clippy::unit_arg)] ::yew::virtual_dom::VChild<Foo>::new(...);
```

and the compiler really doesn't like that.

Reply:
Ah ok, no worries! Thanks for checking 👍